testing/testPlan.md

```diff
@@ -0,0 +1,172 @@
+
+```
+import unittest
+
+class TestStringMethods(unittest.TestCase):
+
+    def test_upper(self):
+        self.assertEqual('foo'.upper(), 'FOO')
+
+    def test_isupper(self):
+        self.assertTrue('FOO'.isupper())
+        self.assertFalse('Foo'.isupper())
+
+    def test_split(self):
+        s = 'hello world'
+        self.assertEqual(s.split(), ['hello', 'world'])
+        # check that s.split fails when the separator is not a string
+        with self.assertRaises(TypeError):
+            s.split(2)
+
+if __name__ == '__main__':
+    unittest.main()
+```
+
+The first test `test_upper()` demonstrates how simple it is to create a unit 
+test. This test requires no setup or teardown, only the UUT which is the 
+`.upper` function of the string class. `test_split()` is a test that shows a 
+setup, the `s = 'hello world'`, and also shows a demonstration of negative 
+testing. Negative testing is when we expect the UUT to NOT do something. An 
+example of this is seen in the above code where we expect that `s.split()` to 
+fail when the parameter to `s.split()` is the int 2, and not a string.
+
+The following is an example using a function `TraceFileParser.open` from our 
+from our codebase.
+
+```
+def open(self) -> bool: 
+    """Open the trace file""" 
+    try: 
+        self.fd = open(self.filename, 'r', encoding="utf-8") 
+        return True
+    except FileNotFoundError:    
+        print(f"[ERROR] - could not find trace file '{self.filename}'") 
+        return False
+    except PermissionError: 
+        print(f"[ERROR] - Permission denied accessing file: '{self.filename}'")
+        return False
+    except Exception as e: 
+        print(f"[ERROR] - Opening file: {e}")
+        return False
+```
+
+A unit test for this function could looks something like this:
+
+```
+class traceFileParser_Open(unittest.TestCase)
+
+    def test_openFile(self):
+        parser = TraceFileParser()
+        self.assertTrue(parser.open("filename"))
+    
+    def test_FileNotFoundError:
+        parser = TraceFileParser()
+        self.assertFalse(parser.open("bad_filename"))
+
+    def test_FileNotFoundError
+        parser = TraceFileParser()
+        self.assertFalse(parser.open("file_with_bad_permissions))
+
+    def test_FileNotFoundError
+        parser = TraceFileParser()
+        self.assertFalse(parser.open("file_with_exception"))
+```
+
+Note, this would be a good function to refactor for returning the error instead
+of immediately printing it. Then we can test the error message as well.
+
+[Link to more unit test examples](https://github.com/mdhardenburgh/matrixLibrary/blob/main/unitTest/matrixTest.cpp).
+These examples are in C++, but you can see the 2 of the 3 parts of a unit test: 
+setup and verification. These examples did not need the optional teardown. 
+
+#### Note on functions that do not return a value
+These are a bit harder to test. In my experience, the function is modified to
```

Comment:
We could also spy on the side-effects in most cases, probably using a mock with a method that would assert the side effect function was called with a particular value, or called a specific number of times.

Reply:
I'm currently contemplating this same conundrum with the logging utility. Test plan looks good to me though.  
![ucvzC0l_d](https://github.com/user-attachments/assets/dc6d9c53-f832-4c1a-9b39-d5a28266f764)

Reply:
Yeah, I think we will probably want to do a little bit of both - try to think about how to write functions that are easily testable, but also be able to check for side effects using testing capabilities like checking whether a function was called.